File: src/main/java/org/example/kalkulationsprogramm/dto/Verrechnungslohn/VerrechnungslohnErgebnisDto.java
package org.example.kalkulationsprogramm.dto.Verrechnungslohn;

import lombok.Getter;
import lombok.Setter;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

@Getter
@Setter
public class VerrechnungslohnErgebnisDto {

    public enum Modus {
        RUECKWIRKEND,
        HOCHRECHNUNG
    }

    public enum LohnQuelle {
        LOHNABRECHNUNG,
        KALKULATORISCH,
        STUNDENLOHN_HOCHRECHNUNG,
        STAMMSTUNDENLOHN
    }

    private int jahr;
    private Modus modus;

    private List<MitarbeiterLohnZeile> lohnzeilen = new ArrayList<>();
    private List<MitarbeiterStundenZeile> stundenzeilen = new ArrayList<>();
    private List<KostenstelleAnteil> kostenstellen = new ArrayList<>();
    private List<AbteilungVorschlag> abteilungen = new ArrayList<>();
    private List<DatenLuecke> datenLuecken = new ArrayList<>();

    private BigDecimal lohnsummeGesamt = BigDecimal.ZERO;
    private BigDecimal verkaeuflicheStundenGesamt = BigDecimal.ZERO;
    private BigDecimal gemeinkostenGesamt = BigDecimal.ZERO;
    private BigDecimal selbstkostenProStunde = BigDecimal.ZERO;

    @Getter
    @Setter
    public static class MitarbeiterLohnZeile {
        private Long mitarbeiterId;
        private String name;
        private boolean istGeschaeftsfuehrer;
        private String beschaeftigungsart;
        private BigDecimal bruttoJahr = BigDecimal.ZERO;
        private BigDecimal agAnteilSv = BigDecimal.ZERO;
        private BigDecimal bgBeitrag = BigDecimal.ZERO;
        private BigDecimal geldwerterVorteilJahr = BigDecimal.ZERO;
        private BigDecimal gesamtkosten = BigDecimal.ZERO;
        private LohnQuelle quelle;
        private boolean bruttoIstDefault;
    }

    @Getter
    @Setter
    public static class MitarbeiterStundenZeile {
        private Long mitarbeiterId;
        private String name;
        private boolean istGeschaeftsfuehrer;
        private BigDecimal sollstunden = BigDecimal.ZERO;
        private BigDecimal urlaubsstunden = BigDecimal.ZERO;
        private BigDecimal krankheitsstunden = BigDecimal.ZERO;
        private BigDecimal interneStunden = BigDecimal.ZERO;
        private BigDecimal feiertagsstunden = BigDecimal.ZERO;
        private BigDecimal verkaeuflicheStunden = BigDecimal.ZERO;
        private boolean urlaubIstDefault;
        private boolean krankheitIstDefault;
        private boolean interneIstDefault;
    }

    @Getter
    @Setter
    public static class KostenstelleAnteil {
        private Long kostenstelleId;
        private String bezeichnung;
        private BigDecimal jahresbetrag = BigDecimal.ZERO;
        private boolean gestreckt;
    }

    @Getter
    @Setter
    public static class AbteilungVorschlag {
        private Long abteilungId;
        private String name;
        private BigDecimal aufschlagEuro = BigDecimal.ZERO;
    }

    @Getter
    @Setter
    public static class DatenLuecke {
        private Long mitarbeiterId;
        private String mitarbeiterName;
        private String problem;
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/dto/Verrechnungslohn/VerrechnungslohnErgebnisDto.kt
package org.example.kalkulationsprogramm.dto.Verrechnungslohn

import java.math.BigDecimal

class VerrechnungslohnErgebnisDto {
    enum class Modus { RUECKWIRKEND, HOCHRECHNUNG }
    enum class LohnQuelle { LOHNABRECHNUNG, KALKULATORISCH, STUNDENLOHN_HOCHRECHNUNG, STAMMSTUNDENLOHN }

    var jahr: Int = 0
    var modus: Modus? = null
    var lohnzeilen: MutableList<MitarbeiterLohnZeile> = ArrayList()
    var stundenzeilen: MutableList<MitarbeiterStundenZeile> = ArrayList()
    var kostenstellen: MutableList<KostenstelleAnteil> = ArrayList()
    var abteilungen: MutableList<AbteilungVorschlag> = ArrayList()
    var datenLuecken: MutableList<DatenLuecke> = ArrayList()
    var lohnsummeGesamt: BigDecimal = BigDecimal.ZERO
    var verkaeuflicheStundenGesamt: BigDecimal = BigDecimal.ZERO
    var gemeinkostenGesamt: BigDecimal = BigDecimal.ZERO
    var selbstkostenProStunde: BigDecimal = BigDecimal.ZERO

    data class MitarbeiterLohnZeile(
        var mitarbeiterId: Long? = null,
        var name: String? = null,
        var isIstGeschaeftsfuehrer: Boolean = false,
        var beschaeftigungsart: String? = null,
        var bruttoJahr: BigDecimal = BigDecimal.ZERO,
        var agAnteilSv: BigDecimal = BigDecimal.ZERO,
        var bgBeitrag: BigDecimal = BigDecimal.ZERO,
        var geldwerterVorteilJahr: BigDecimal = BigDecimal.ZERO,
        var gesamtkosten: BigDecimal = BigDecimal.ZERO,
        var quelle: LohnQuelle? = null,
        var isBruttoIstDefault: Boolean = false,
    )
    data class MitarbeiterStundenZeile(
        var mitarbeiterId: Long? = null,
        var name: String? = null,
        var isIstGeschaeftsfuehrer: Boolean = false,
        var sollstunden: BigDecimal = BigDecimal.ZERO,
        var urlaubsstunden: BigDecimal = BigDecimal.ZERO,
        var krankheitsstunden: BigDecimal = BigDecimal.ZERO,
        var interneStunden: BigDecimal = BigDecimal.ZERO,
        var feiertagsstunden: BigDecimal = BigDecimal.ZERO,
        var verkaeuflicheStunden: BigDecimal = BigDecimal.ZERO,
        var isUrlaubIstDefault: Boolean = false,
        var isKrankheitIstDefault: Boolean = false,
        var isInterneIstDefault: Boolean = false,
    )
    data class KostenstelleAnteil(var kostenstelleId: Long? = null, var bezeichnung: String? = null, var jahresbetrag: BigDecimal = BigDecimal.ZERO, var isGestreckt: Boolean = false)
    data class AbteilungVorschlag(var abteilungId: Long? = null, var name: String? = null, var aufschlagEuro: BigDecimal = BigDecimal.ZERO)
    data class DatenLuecke(var mitarbeiterId: Long? = null, var mitarbeiterName: String? = null, var problem: String? = null)
}
